execution.go

```diff
@@ -222,27 +207,73 @@ func (s *ExecutableSchema) ExecuteQuery(ctx context.Context) *graphql.Response {
 		}
 	}
 
-	for name, value := range extensions {
-		graphql.RegisterExtension(ctx, name, value)
+	for _, result := range results {
+		errs = append(errs, result.Errors...)
 	}
 
-	res, err := marshalResult(result, op.SelectionSet, s.MergedSchema, &ast.Type{NamedType: strings.Title(string(op.Operation))})
+	introspectionData := s.resolveIntrospectionFields(ctx, op.SelectionSet, filteredSchema)
+	if len(introspectionData) > 0 {
+		results = append([]executionResult{
+			{
+				ServiceURL: internalServiceName,
+				Data:       introspectionData,
+			},
+		}, results...)
+	}
+
+	mergeStart := time.Now()
+	mergedResult, err := mergeExecutionResults(results)
 	if err != nil {
 		errs = append(errs, &gqlerror.Error{Message: err.Error()})
 		AddField(ctx, "errors", errs)
 		return &graphql.Response{
 			Errors: errs,
 		}
 	}
+	timings["merge"] = time.Since(mergeStart).Round(time.Millisecond).String()
+
+	bubbleErrs, err := bubbleUpNullValuesInPlace(qe.schema, op.SelectionSet, mergedResult)
+	if err == errNullBubbledToRoot {
+		mergedResult = nil
+	} else if err != nil {
+		errs = append(errs, &gqlerror.Error{Message: err.Error()})
+		AddField(ctx, "errors", errs)
+		return &graphql.Response{
+			Errors: errs,
+		}
+	}
+
+	errs = append(errs, bubbleErrs...)
+
+	formattingStart := time.Now()
+	formattedResponse, err := formatResponseData(qe.schema, op.SelectionSet, mergedResult)
+	if err != nil {
+		errs = append(errs, &gqlerror.Error{Message: err.Error()})
+		AddField(ctx, "errors", errs)
+		return &graphql.Response{
+			Errors: errs,
+		}
+	}
+	timings["format"] = time.Since(formattingStart).Round(time.Millisecond).String()
+
+	if debugInfo, ok := ctx.Value(DebugKey).(DebugInfo); ok {
+		if debugInfo.Timing {
+			extensions["timing"] = timings
+		}
+	}
+
+	for name, value := range extensions {
+		graphql.RegisterExtension(ctx, name, value)
+	}
 
 	if len(errs) > 0 {
 		AddField(ctx, "errors", errs)
 	}
+
 	return &graphql.Response{
-		Data:   res,
+		Data:   []byte(formattedResponse),
```

Comment:
Is this going to create large copy calls when casting a `string` to `[]byte`? It doesn't seem that in `formatResponseData` the type is required to be a string, you could build up a `[]byte` buffer instead.